New module: keycloak_group

adamgoossens opened the pull request:
##### SUMMARY
This module (keycloak_group) allows using the Keycloak REST API (http://www.keycloak.org/docs-api/3.4/rest-api/index.html) to administrate Keycloak (http://www.keycloak.org) groups. It allows users to create, update and remove groups.

Keycloak is an Open Source Identity and Access Management system spearheaded by Red Hat. It provides OpenID Connect and SAML authentication/authorization services.

Two prior modules in this namespace have been merged (keycloak_client in #31716 and keycloak_clienttemplate in #33419). This module reuses and extends code introduced therein (from module_utils/keycloak.py).

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
keycloak_group

##### ANSIBLE VERSION
2.8.0

adamgoossens committed:
Initial implementation of groups module.

Not all parameters are supported, yet.

adamgoossens committed:
Clarify read-only status of realmRoles, clientRoles and access attributes

adamgoossens committed:
Fix testing failures

adamgoossens committed:
Fix additional style issues.

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+            - State of the group.
+            - On C(present), the group will be created if it does not yet exist.
+            - On C(absent), the group will be removed if it exists.
+            - On C(updated), the group will be updated or created based on the parameters you provide.
```
This seems to be a leftover, there is no state "updated" anywhere else in the code.

adamgoossens replied:
....yep, it certainly is. Will fix :)

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+        description:
+            - Name of the group.
+            - This parameter is required only when creating or updating the group.
+        required: false
```
When required is false, it can be left out (since it is the default).

adamgoossens replied:
Ah, of course - thanks!

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+    attributes:
+        description:
+            - A list of key/value pairs to set as custom attributes for the group.
+            - Values may be single values (e.g. a string) or a list of strings.
```
Are there any attributes commonly available and with defined semantics in, for instance, Keycloak 3.4, or is this completely freeform user-definable?

adamgoossens replied:
Totally user-definable key value pairs for storing additional custom data on the group.

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+
+notes:
+    - Presently, the I(realmRoles), I(clientRoles) and I(access) attributes returned by the Keycloak API
+      are read-only for groups. This limitation will be removed in a later version of this module.
```
This should be fine; if and when you want to implement this, please have a look at the check_role_representation in #35558, it might be helpful :)

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+    auth_realm: master
+    auth_username: USERNAME
+    auth_password: PASSWORD
+  register: kc_group_id
```
Does this register do anything, i.e. is it used somewhere?

adamgoossens replied:
I've used it in test playbooks to retrieve the auto-generated group ID. I'll remove it since it doesn't really value add to the example.

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+            module.params['attributes'][key] = [val] if not isinstance(val, list) else val
+
+    group_params = [x for x in module.params
+                    if x not in list(keycloak_argument_spec().keys()) + ['state', 'realm'] and
```
I will steal this.

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+
+            module.exit_json(**result)
+
+        else:
```
This branch can never be hit since the argument_spec defines which states are valid and checks the provided module arguments against it.

adamgoossens replied:
Good point - just me being overcautious. Will remove.

eikef commented on lib/ansible/module_utils/keycloak.py:
```diff
@@ -339,3 +341,134 @@ def delete_client_template(self, id, realm="master"):
+
+        If the group does not exist, None is returned.
+
+        GID is a UUID provided by the Keycloak API
```
probably lowercase gid here :)

eikef reviewed (changes requested):
Code looks good, with some minor changes.

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+            module.fail_json(msg='name must be specified when creating a new group')
+
+        if module._diff:
+            result['diff'] = dict(before='', after=after_group)
```
after_group does not exist at this point, this will stacktrace

eikef commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,358 @@
+    auth_password: PASSWORD
+
+- name: Delete a Keycloak group based on name
+  local_action:
```
This is missing a state, so the group is created instead of being deleted.

eikef commented:
The module does what it is supposed to do -- creation, deletion, and updates of groups all work. There are some minor code issues, once they are addressed this looks good to me to ship.

adamgoossens committed:
Minor updates and fixes after review feedback

adamgoossens commented:
ready_for_review

eikef commented:
shipit

getjack commented:
This PR seems to be stale. Can we still shipit?

RickS-C137 commented:
[shipit]

adamgoossens commented:
The code itself should still be good, but I will need to amend the version tag - currently marked as available from Ansible 2.5 :)

resmo commented:
I would like to merge this into 2.8. 

However, I don't like the the "return". It returns 3 states of the same thing: wanted, existing, end_stateTo me this looks over complicated. I would suggest to only return the "real values" and document them properly.

This way, integration tests can easily be done as we do in other modules:

- create in check mode --> expect empty dict return and changed=true
- create in live mode --> return dict with values and changed=true
- create in live mode for idempotency test --> return dict with values and changed=false

See https://github.com/ansible/ansible/blob/devel/test/integration/targets/vultr_user/tasks/main.yml as an example. Any thoughts?

resmo commented:
needs_info

adamgoossens commented:
Hi @resmo , I have no problems with that. I will spend some time on this this weekend and get it ready for review again in time for 2.8.

adamgoossens committed:
Simplify return values after feedback

This removes the 'proposed' and 'end_state' return values, replacing them instead with just 'group'. 'group' is the representation of the group after the module completes.

Also update the dates, and the Ansible version

adamgoossens committed:
Corrections after module validation

adamgoossens commented:
ready_for_review

resmo commented:
well done. I only see a few minor cosmetic things in docs which should not block this from merging, I'll comment inline.

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+    meta_args = dict(
+        state=dict(default='present', choices=['present', 'absent']),
+        realm=dict(default='master'),
+
```
space looks unnecessary

adamgoossens replied:
Agreed - removed.

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+                           supports_check_mode=True,
+                           required_one_of=([['id', 'name']]))
+
+    result = dict(changed=False, msg='', diff={}, group_id='')
```
shouldn't this be  group instead of group_id?

~~~diff
 -    result = dict(changed=False, msg='', diff={}, group_id='')
 +    result = dict(changed=False, msg='', diff={}, group='')
~~~

adamgoossens replied:
Agreed - fixed.

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+group:
+  description: Group representation of the group after module execution (sample is truncated).
+  returned: always
+  type: dict
```
this should be type complex and document every possible return in this, see https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/hcloud/hcloud_ssh_key.py#L88 as an example

adamgoossens replied:
This is now done. The Keycloak API isn't particularly well documented, however this particular set of return values has been pretty stable over the last few versions. I don't expect description for this complex return value to change.

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+'''
+
+RETURN = '''
+msg:
```
`msg` is a common return and not module specific and must not be documented. Please remove

adamgoossens replied:
Done. Thanks!

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+    auth_password: PASSWORD
+
+- name: Update the name of a Keycloak group
+  local_action:
```
not a blocker but `local_action` is a thing the ansible_linter suggest to change. That is why I would prefer:

~~~yaml

- name: Update the name of a Keycloak group
  keycloak_group:
    id: '9d59aa76-2755-48c6-b1af-beb70a82c3cd'
    name: an-updated-kc-group-name
    state: present
    auth_client_id: admin-cli
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
  delegate_to: localhost
~~~

adamgoossens replied:
Agreed - I've now changed and pushed this. Previous commit missed the delegate_to option.

resmo commented on lib/ansible/modules/identity/keycloak/keycloak_group.py:
```diff
@@ -0,0 +1,323 @@
+            - On C(present), the group will be created if it does not yet exist, or updated with the parameters you provide.
+            - On C(absent), the group will be removed if it exists.
+        required: true
+        default: 'present'
```
not a blocker but we started to add a `type: (bool|str|dict|int)` for each option as well, this would look like 

~~~diff
         default: 'present'
+        type: str
~~~

adamgoossens replied:
Thanks, I've added types for each of the module options.

resmo reviewed (changes requested):
Please let me know if you shouldn't have time to work on these minors before the freeze on 4.4.19

adamgoossens commented:
> Please let me know if you shouldn't have time to work on these minors before the freeze on 4.4.19

Thanks for the comprehensive review! I'll tackle these tomorrow my time. Should be no problem to have them completed by end of this weekend, ahead of the freeze.

adamgoossens committed:
Further documentation updates after feedback

Minor whitespace adjustments also

adamgoossens committed:
Add delegate_to: localhost stanzas to examples

adamgoossens commented:
ready_for_review

resmo reviewed (approved):
LGTM

adamgoossens commented:
Thanks @resmo - hasn't seem to have merged however; is there anything else I need to do before 04 Apr? @eikef does it also need another tick from you as the maintainer?

resmo commented:
We have a little more time because the freeze has been delayed to 11.04. I would be good if @eikef give a final 'shipit'. However, new modules won't get automerged and I have to merge it. 

I realized that the existing modules in for keycloak also have this "proposed, existing, end_state", I would suggest not to continue this existing return pattern and follow the way you did in this module.

I will merge it, if we don't get a "veto" by @eikef

resmo commented:
rebuild_merge

dazasa commented:
Hi, do you intend to implement the role mapping in this script? Best regards David
